src/video.cpp

```diff
@@ -5,861 +5,908 @@
+	}
+
+	FT_Select_Charmap(face, FT_ENCODING_UNICODE);
+
+	FT_Set_Pixel_Sizes(face, 0, font_size);
+	glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+
+	/* Load all characters we may need. */
+	for (uint32 codepoint = 1; codepoint <= MAX_CODEPOINT; codepoint = NextCodepointToLoad(codepoint)) {
+		if (FT_Load_Char(face, codepoint, FT_LOAD_RENDER) != 0) {
+			this->characters[codepoint].valid = false;
+			char buffer[] = {0, 0, 0, 0, 0};
+			EncodeUtf8Char(codepoint, buffer);
+			printf("WARNING: Failed to load glyph U+%04x '%s'\n", codepoint, buffer);
+			continue;
+		}
+
+		GLuint texture;
+		glGenTextures(1, &texture);
+		glBindTexture(GL_TEXTURE_2D, texture);
+		glTexImage2D(
+				GL_TEXTURE_2D,
+				0,
+				GL_RED,
+				face->glyph->bitmap.width,
+				face->glyph->bitmap.rows,
+				0,
+				GL_RED,
+				GL_UNSIGNED_BYTE,
+				face->glyph->bitmap.buffer
+		);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+
+		this->characters[codepoint] = {
+			texture,
+			Point16(face->glyph->bitmap.width, face->glyph->bitmap.rows),
+			Point16(face->glyph->bitmap_left, face->glyph->bitmap_top),
+			static_cast<GLuint>(face->glyph->advance.x),
+			true
+		};
+	}
+
+	glBindTexture(GL_TEXTURE_2D, 0);
+	FT_Done_Face(face);
+	FT_Done_FreeType(ft);
+
+	/* Check that we have at least a bearing character and a glyph for invalid characters. */
+	std::string sample_text = {BEARING_CHARACTER};
+	const char *c = sample_text.c_str();
+	size_t i = 1;
+	this->GetFontGlyph(&c, i);
+	this->GetFontGlyph(&c, i);
 }
 
 /**
- * Copy constructor.
- * @param cr Existing clipped rectangle.
+ * Look up the font glygh to use for a given character.
+ * If the current font does not have a matching glygh, a default value is returned.
+ * @param text [inout] Pointer to the UTF-8 encoded character. Will be advanced to the next character.
+ * @param length [inout] Number of bytes left in the text. Will be decremented by the number of bytes by which the text is advanced.
+ * @return Glyph to use.
  */
-ClippedRectangle::ClippedRectangle(const ClippedRectangle &cr)
+const TextRenderer::FontGlyph &TextRenderer::GetFontGlyph(const char **text, size_t &length) const
 {
-	this->absx = cr.absx;
-	this->absy = cr.absy;
-	this->width = cr.width;
-	this->height = cr.height;
-	this->address = cr.address; this->pitch = cr.pitch;
+	uint32 codepoint;
+	int bytes_read = length < 1 ? 0 : DecodeUtf8Char(*text, length, &codepoint);
+
+	if (bytes_read < 1) {
+		if (length > 0) {
+			++*text;
+			++length;
+		}
+		/* Fall though to default glyph selection. */
+	} else {
+		*text += bytes_read;
+		length -= bytes_read;
+		if (codepoint <= MAX_CODEPOINT && this->characters[codepoint].valid) return this->characters[codepoint];
+		/* Fall though to default glyph selection. */
```

Comment:
Insert an empty line before or write `/* else fall though to default glyph selection. */` to make it more obvious that this doesn't happen always.